pkg/strategy/dca2/recover.go

```diff
@@ -0,0 +1,266 @@
+package dca2
+
+import (
+	"context"
+	"fmt"
+	"sort"
+	"strconv"
+	"time"
+
+	"github.com/c9s/bbgo/pkg/bbgo"
+	"github.com/c9s/bbgo/pkg/core"
+	"github.com/c9s/bbgo/pkg/fixedpoint"
+	"github.com/c9s/bbgo/pkg/types"
+)
+
+type queryAPI interface {
+	QueryOpenOrders(ctx context.Context, symbol string) ([]types.Order, error)
+	QueryClosedOrdersDesc(ctx context.Context, symbol string, since, until time.Time, lastOrderID uint64) ([]types.Order, error)
+	QueryOrderTrades(ctx context.Context, q types.OrderQuery) ([]types.Trade, error)
+}
+
+func (s *Strategy) recover(ctx context.Context) error {
+	s.logger.Info("[DCA] recover")
+	queryService, ok := s.Session.Exchange.(queryAPI)
+	if !ok {
+		return fmt.Errorf("[DCA] exchange %s doesn't support queryAPI interface", s.Session.ExchangeName)
+	}
+
+	openOrders, err := queryService.QueryOpenOrders(ctx, s.Symbol)
+	if err != nil {
+		return err
+	}
+
+	closedOrders, err := queryService.QueryClosedOrdersDesc(ctx, s.Symbol, time.Time{}, time.Now(), 0)
+	if err != nil {
+		return err
+	}
+
+	currentRound, err := getCurrentRoundOrders(s.Short, openOrders, closedOrders, s.OrderGroupID)
+	if err != nil {
+		return err
+	}
+	debugRoundOrders(s.logger, "current", currentRound)
+
+	// recover state
+	state, err := recoverState(ctx, s.Symbol, s.Short, int(s.MaxOrderNum), openOrders, currentRound, s.OrderExecutor.ActiveMakerOrders(), s.OrderExecutor.OrderStore(), s.OrderGroupID)
+	if err != nil {
+		return err
+	}
+
+	// recover position
+	if err := recoverPosition(ctx, s.Position, queryService, currentRound); err != nil {
+		return err
+	}
+
+	// recover budget
+	budget := recoverBudget(currentRound)
+
+	// recover startTimeOfNextRound
+	startTimeOfNextRound := recoverStartTimeOfNextRound(ctx, currentRound, s.CoolDownInterval)
+
+	s.state = state
+	if !budget.IsZero() {
+		s.Budget = budget
+	}
+	s.startTimeOfNextRound = startTimeOfNextRound
+
+	return nil
+}
+
+// recover state
+func recoverState(ctx context.Context, symbol string, short bool, maxOrderNum int, openOrders []types.Order, currentRound Round, activeOrderBook *bbgo.ActiveOrderBook, orderStore *core.OrderStore, groupID uint32) (State, error) {
+	numOpenOrders := len(openOrders)
+	// dca stop at take profit order stage
+	if currentRound.TakeProfitOrder.OrderID != 0 {
+		// check the open orders is take profit order or not
+		if numOpenOrders == 1 {
+			if openOrders[0].OrderID == currentRound.TakeProfitOrder.OrderID {
+				activeOrderBook.Add(openOrders[0])
+				// current round's take-profit order still opened, wait to fill
+				return TakeProfitReady, nil
+			} else {
+				return None, fmt.Errorf("stop at taking profit stage, but the open order's OrderID is not the take-profit order's OrderID")
+			}
+		}
+
+		if numOpenOrders == 0 {
```

Comment:
you can check numOpenOrders == 0 before you check numOpenOrders == 1 ? so we can do early return here?

Reply:
OK!